### DISCUSSION TITLE
How to export the chat history of GitHub Copilot Chat?

### DISCUSSION THREAD

**Discussion by z14git** [community/community#57190] [Copilot Conversations]
### Select Topic Area

Question

### Body

Currently, I am using Copilot Chat to assist in the development of a repository's code. While I can easily synchronize the code to other computers, I cannot do the same for the chat history of Copilot Chat. This inconvenience me when developing on other computers. I hope that in the future, there will be an added feature to export and import chat history.

**Comment by robvet** [comment]
Agree this would be helpful feature. 

There come a time when you want to come back to earlier Copilot chat entries.

A way to save select output would be value add.

**Reply by joshuadanpeterson** [reply]
You still can see the chat history, but yeah, an export option would be nice. I asked it about exporting and it said to take a screenshot or to copy and paste the conversation to a text file. I keep a journal for programming, and I've been taking advantage of ChatGPT's individual URLs for conversation threads to refer back to various queries and it'd be great if there was a similar feature in Copilot Chat

**Reply by ChristianOConnor** [reply]
How do I get old chat logs? Sometimes my chats disappear. Where are they stored? For example, on a mac I noticed there's a file called "Code - Insiders" in ~/Library/Application Support. Are the chats saved here?

**Reply by AvremiB** [reply]
In VSC, you can see the HTML of a chat displayed in the VSC Dev Tools, then copy and save the HTML. I haven't tried it (and you should note that references to CSS files may be relative) but in theory it seems like a good idea.

**Comment by Marijn-Bergman** [comment]
You can export a Copilot Chat session in Visual Studio Code to a JSON file via `Ctrl+Shift+P | Chat: Export Session...` or by choosing `View > Command Palette | Chat: Export Session...`
You can then import this chat.json file using `Ctrl+Shift+P | Chat: Import Session...` or by choosing `View > Command Palette | Chat: Import Session...`

If you would like to read your chat session outside of VSC i made a [python script](https://github.com/Marijn-Bergman/copilot-chat-export-formatter) that formats the .json file and puts a more readable version in a .txt file.

**Reply by ChristianOConnor** [reply]
This is very helpful thanks! Also, do you know if there's a way to recover Github Copilot Chat sessions that have disappeared for whatever reason? For example, if you delete or move a directory in which you had a Github Copilot chat going?

**Reply by tamvcspk** [reply]
You saved my day

**Reply by maximojoexpress** [reply]
FYI import failed on a recently exported chat.

**Reply by rhlobo** [reply]
Thank you for the script. It did not work, I guess the json structure has changed, but it served as inspiration for me to stop being lazy and doing what I needed: I've created an [script that creates an markdown from the chat json](https://gist.github.com/rhlobo/4f7f42058d285d48b96c80e0cf042dd1). Hope it helps someone...

**Reply by stufflebeam** [reply]
> Hope it helps someone...

It did!

This is awesome, and exactly what I needed today. So, it is _very much_ appreciated. Thank you!

**Reply by TryTryAgain** [reply]
I see the python provided above is nice to have, I only saw that after I used copilot to help come up with this that I find is pretty nice too:
`jq -r '.requests[] | "Request:\n" + .message.text + "\n\nResponse:\n" + .response[0].value + "\n"' chat-export.json > chat-export.md`

**Reply by Di9** [reply]
> I see the python provided above is nice to have, I only saw that after I used copilot to help come up with this that I find is pretty nice too: `jq -r '.requests[] | "Request:\n" + .message.text + "\n\nResponse:\n" + .response[0].value + "\n"' chat-export.json > chat-export.md`

this is the best solutions! just write it as main (separate) comment, so it could be upvote!

**Reply by CoyoteLeo** [reply]
If you are using [open-webui](https://github.com/open-webui/open-webui), you can use this [script](https://gist.github.com/CoyoteLeo/47ed7e29d29b8b24e1c875b02a6d0859) to transform the exported json to the foramt that open-webui accepted.

**Reply by laggingreflex** [reply]
Does import not work or has stopped working? (v1.99.0, chat:0.26.0) 

Export is ok (all details are there in the json), but importing shows nearly empty chat (just shows the "Used 1 reference" widget).

**Reply by xpmatteo-tw** [reply]
I got Copilot to write a json-to-markdown python script in no time! We live in amazing times

**Reply by ElectricRCAircraftGuy** [reply]
@xpmatteo-tw , please share the script.

**Reply by xpmatteo-tw** [reply]
It's nothing fancy: "please write a python script that parses a file like #file:refactor-handlers-chat.json and produces a markdown file".  Since then I learned that you can just right-click on the chat sidebar and select "Copy All": it will copy the whole content of the chat as markdown.  But the markdown produced by the script is nicer :)

**Reply by gregce** [reply]
Hey all (specifically @ElectricRCAircraftGuy) In the case its helpful made a VSCode extension for GitHub Copilot that allows your to export chat history, share it and it automatically builds copilot instructions for you, check it out and let us know what you think: https://marketplace.visualstudio.com/items?itemName=SpecStory.specstory-vscode. 

It also autosaves everything into a `.specstory` folder in your project folder.

**Reply by berkaykrc** [reply]
There is `Chat: Import Chat...` and `Chat: Export Chat...` options but they do not stay with you so whenever you need them you have to import again. I did not see Chat: Export Session option on command palette.

**Reply by sascharo** [reply]
> You can export a Copilot Chat session in Visual Studio Code to a JSON file via `Ctrl+Shift+P | Chat: Export Session...` or by choosing `View > Command Palette | Chat: Export Session...`

Was this removed in later version? I do not see that in 1.102.0-Insider.

**Reply by berkaykrc** [reply]
> > You can export a Copilot Chat session in Visual Studio Code to a JSON file via `Ctrl+Shift+P | Chat: Export Session...` or by choosing `View > Command Palette | Chat: Export Session...`
> 
> Was this removed in later version? I do not see that in 1.102.0-Insider.

I have no clue. Using 1.101.0 version from Fedora dnf manager

**Comment by miron** [comment]
> ~/Library/Application Support. Are the chats



> This is very helpful thanks! Also, do you know if there's a way to recover Github Copilot Chat sessions that have disappeared for whatever reason? For example, if you delete or move a directory in which you had a Github Copilot chat going?

In the chat window on top click this icon ![image](https://github.com/community/community/assets/109377/435e8182-2e8d-48d0-baee-90f37ca0f25b), then select your session in the dropdown.
If your chat window does not have this icon, click on the copilot icon ![image](https://github.com/community/community/assets/109377/9340f864-01a5-4473-8274-86410147eb2e) on the bottom right, and select ![image](https://github.com/community/community/assets/109377/539f54a0-28f7-44e8-b14f-4a9ec0a1a4ec) from the search bar. This one has it.

**Reply by joshuadanpeterson** [reply]
That wasn't available when this question was asked.

I'm currently accessing the icon, but no chat sessions are showing even though I use it pretty regularly

**Reply by marko-hologram** [reply]
@joshuadanpeterson yeah it's weird how it behaves. I tried closing vscode now and it preserved a chat session I had started yesterday, but for some reason it doesn't have some some other previous sessions from few days ago. Would love to have a proper history.

**Comment by tonywoode** [comment]
> You can export a Copilot Chat session in Visual Studio Code to a JSON file via `Ctrl+Shift+P | Chat: Export Session...` or by choosing `View > Command Palette | Chat: Export Session...` You can then import this chat.json file using `Ctrl+Shift+P | Chat: Import Session...` or by choosing `View > Command Palette | Chat: Import Session...`
> 

For me, exporting and then importing the immediate result ends in `The editor could not be opened due to an unexpected error: i.with is not a function`

It would seem to be quite important in the long term for a more automated method of keeping keep chat history with the resulting code, in the shorter term a way of being more comfortable that your chats aren't liable to be lost due to some action/inaction

**Reply by james-berkheimer** [reply]
Hi, I just ran into this same error myself.  Did you ever resolve your issue?  Or is this a persistent issue?

**Reply by gregce** [reply]
Hey all (@tonywoode  In the case its helpful made a VSCode extension for GitHub Copilot that allows your to export chat history, share it and it automatically builds copilot instructions for you, check it out and let us know what you think: https://marketplace.visualstudio.com/items?itemName=SpecStory.specstory-vscode. 

It also autosaves everything into a `.specstory` folder in your project folder.

**Comment by ElectricRCAircraftGuy** [comment]
To export your chats to JSON or Markdown, see my comment here: https://github.com/microsoft/vscode-copilot-release/issues/417#issuecomment-1732650912:

> ## How to save GitHub Copilot chats into JSON or Markdown format
>
> To help others verify and duplicate the problem, please post your entire conversation, in JSON format, like this: with your chat open, press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd> to open the command palette, type in `Chat: Export Session...`, save the `chat.json` file. Drag and drop it here.
> 
> You can also get a nice Markdown version by right-clicking in the chat and going to "Copy All".

**Reply by SnowyYANG** [reply]
what about visual studio? I am developing a WPF application with github copilot chat.

**Reply by ElectricRCAircraftGuy** [reply]
I don't know anything about Microsoft Visual Studio. I've never seen nor used it.

**Reply by TryTryAgain** [reply]
ha, the right-clicking and Copy All is very nice, can't believe I didn't see that... it's almost identical to what I was using and posted above as `jq -r '.requests[] | "Request:\n" + .message.text + "\n\nResponse:\n" + .response[0].value + "\n" chat-session.json > chat-session.md'` except the Request=YourUserName and Response=GitHub Copilot ... that's funny

**Reply by abrahampe** [reply]
@rhlobo  This script saved my life, just giving my two cents, I have replace the file opening lines (added utf-8 enconding), due to an error on loading saved json file: 
with open(filename, 'r', encoding="utf-8") as file:
with open(f'{filename}.md', 'w', encoding="utf-8") as file:

**Reply by sascharo** [reply]
> To help others verify and duplicate the problem, please post your entire conversation, in JSON format, like this: with your chat open, press Ctrl + Shift + P to open the command palette, type in `Chat: Export Session...`, save the `chat.json` file. Drag and drop it here.

I do not see this command in the command palette anymore. Has it been removed?

**Comment by iyusuf** [comment]
Right click --> Copy All
Create a new file, paste the clipboard content, save as a markdown file. 
Working great for me.

**Reply by proeslerucsd** [reply]
This is the best answer after attempting scripting, and other printing, or copy paste methods!

**Comment by KhalfaniW** [comment]
Each workspace has its own chat history collection.  You can extract the chats for one workspace using a script like below in the workspace dir with the `Code\User\workspaceStorage\<alphanumerical>\state.vscdb` file.

This will create a group of files in the dir that is similar to the json you would get if you ran `Chat: Export Chat...` command in vscode. It just with a few extra properties. 


``` bash
#!/bin/bash
#extract the history to json
sqlite3 ./state.vscdb "SELECT value FROM ItemTable WHERE key = 'interactive.sessions';" > interactive.sessions.json
# Read the JSON file and split each object into a separate file
jq -c '.[]' interactive.sessions.json | while read -r element; do
    # Extract the customTitle and sanitize it for use as a filename
    customTitle=$(echo $element | jq -r '.customTitle' | tr -cd '[:alnum:][:space:]._-' | tr ' ' '_')
    # Ensure the filename is not empty
    if [ -z "$customTitle" ]; then
        customTitle="untitled"
    fi
    
    # Write the element to a file named after the sanitized customTitle
    echo "$element" | jq '.' > "${customTitle}.json"
done
```

`$ bash extract.sh `

I found the workspaces in `~/.config/Code/User/workspaceStorage/` on linux. 


_________________________
The person that found [this](https://github.com/orgs/community/discussions/69740)  also mentioned a  `memento/interactive-session` properity. I am not sure what is but you can also export it using

`sqlite3 ./state.vscdb "SELECT value FROM ItemTable WHERE key = 'memento/interactive-session';" > memento_interactive-session.json`

**Comment by marcia-schulman** [comment]
Visual Studio 2022 saves the GitHub Copilot chat history to log files in the user's AppData\Local\Temp directory. They are named with a timestamp + "VSGitHubCopilot.chat.log" and may be in a subdirectory of Temp.

What's great about this is it gives you not only the question and answer history, but shows you detail of what Copilot was analyzing to prepare the answer it gave you, and also reveals the initial prompt it is using to be your AI Assistant.  FYI in my logs it shows the prompt as this:

> Role: system

> Content:
> You are a world class AI programming assistant who excels in software development.
> When asked your name, you must respond with "GitHub Copilot".
> Follow the user's requirements carefully & to the letter.
> The user is a proficient software developer working in Visual Studio 2022.
> While the user may have experience in software development, you should not elude to their background. This approach respects the user's expertise without immediately categorizing their profession.
> For questions not related to software development, simply give a reminder that you are an AI programming assistant.
> Follow Microsoft content policies and avoid content that violates copyrights.
> Respond in the following locale: en-US

> Respond in Markdown, for multi-line code, use language-specific markdown code fences.
> 
> Ensure your response is short, impersonal, expertly written and easy to understand.
> Before responding take a deep breath and then work on the user's problem step-by-step.
> Focus on being clear, helpful, and thorough without assuming extensive prior knowledge.

> Generated code should adhere to the existing coding style in the provided context.
> When generating code prefer languages provided in context. If the coding language is unclear fallback to generating code in C#.
> Generate code that can be copy & pasted without modification, i.e. preserve surrounding user code, avoid placeholder comments like "existing code here..." etc. 
> After generating mutated code consider mentioning what specifically was changed and your reasoning if it would help the user.
> 
> The active document or selection is the source code the user is looking at right now and is what they care about.
> Additional Rules: The user has a checkmark selected to automatically include the active document.
> At the end of you response, you should notify the user that you are using the active document because they have the checkmark checked.
> You must also notify them that they can include additional context using **#** references. Typing **#** opens a completion list of available context.

**Reply by TechAurelianMisc2** [reply]
> Visual Studio 2022 saves the GitHub Copilot chat history to log files in the user's AppData\Local\Temp directory. They are named with a timestamp + "VSGitHubCopilot.chat.log" and may be in a subdirectory of Temp.

@marcia-schulman Do you happen to know where / if Visual Studio Code saves the GitHub Copilot chat history on macOS (and/or Windows)?

**Reply by sascharo** [reply]
> Do you happen to know where / if Visual Studio Code saves the GitHub Copilot chat history on macOS (and/or Windows)?

Did you figure it out?

**Reply by iammohitsakhuja** [reply]
For macOS, it seems to be stored in `$HOME/Library/Application\ Support/Code/User/workspaceStorage/<workspace_id>` for each workspace.

**Comment by ParikhKadam** [comment]
Chat histories are stored at `~/.config/Code/User/workspaceStorage/<workspace_id>` in Linux

**Comment by oop7** [comment]
For Windows: `%APPDATA%\Code\User\workspaceStorage\`
For macOS: `~/Library/Application Support/Code/User/workspaceStorage/`
For Linux: `~/.config/Code/User/workspaceStorage/`

Each folder contains chat history for a specific workspace, identified by unique workspace IDs.

**Comment by Fzzzhan** [comment]
Try this, https://marketplace.visualstudio.com/items?itemName=fengzehan.vscode-copilot-exporter. :)

**Comment by shdwkl** [comment]
Try this https://marketplace.visualstudio.com/items?itemName=imperium-dev.copilot-chat-to-markdown 
supporots new format `jsonl`

**Comment by NabiKAZ** [comment]
## Export Copilot Chat to Styled HTML (Quick Guide)

1. **Copy chat**
   - Right-click Copilot chat history → **Copy all**

2. **Prepare Markdown**
   - Paste into a `.md` file
   - Replace:
     ```
     User:
     ```
     with:
     ```
     > User:
     ```
   - For multiline messages, ensure each line starts with `>`

3. **Convert to HTML**
   - Go to https://dillinger.io/
   - Paste Markdown
   - Click **Export → Styled HTML**

4. **Add this at the top of the HTML file**
```html
<style>
* { font-family: vazirmatn, vazir; direction:rtl; }
blockquote { background: #0084ff; color: #ffffff; padding: 10px 14px; margin: 8px 0 8px auto; border-radius: 14px; max-width: 75%; line-height: 1.5; word-wrap: break-word; }
blockquote p { border: none !important; font-style: normal !important; }ont-style: normal !important; }
</style>
```

✅ Open the HTML file in your browser and enjoy it!
